core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -521,13 +551,20 @@ public HttpResponse convertResponse(ServiceRequestContext ctx,
             }
 
             assert f != null;
-            return HttpResponse.from(f.thenApply(aggregated -> {
+            final CompletableFuture<HttpResponse> resFuture = f.thenApply(aggregated -> {
                 try {
                     return responseConverter.convertResponse(ctx, headers, aggregated, trailers);
                 } catch (Exception ex) {
                     return Exceptions.throwUnsafely(ex);
                 }
-            }));
+            });
+            resFuture.handle((ignored, cause) -> {
+                if (cause != null && !f.isDone()) {
+                    f.completeExceptionally(cause);
+                }
+                return null;
+            });
```

Comment:
This might not catch an exception in `HttpResponse`
How about moving this into `f.thenApply(...)`?
```java
try {
    HttpResponse response = responseConverter.convertResponse(ctx, headers, aggregated, trailers);
    CompletableFutureUtil.propagateException(response.whenComplete(), f);
} catch (Exception ex) {
    if(!f.isDone()) {
        f.completeExceptionally(cause);
    }
}
```

Reply:
Wouldn't [this change](https://github.com/line/armeria/pull/3691/files/4bd4e6327e6e0d28fb674c058e8340064b918832#diff-379aaf6327a539eea91e32b7f5393ce200dbc155f30edb6d04486140ecf3cad6R75-R87) in `DeferredHttpResponse` propagate exception from `HttpResponse`?

The propagation chain will be..:
`HttpResponse(DeferredHttpResponse)` -> `resFuture` -> `f` I guess?

Reply:
> The propagation chain will be..:

You are right. 👍 I missed that point.

Reply:
And it seems moving the exception (back)propagation logic into `f.thenApply(...)` will not work since `thenApply(...)` callback is only triggered after `f` is completed but we are trying to complete `f` earlier before it would otherwise have completed.

Reply:
I can't help but agree.

Reply:
By the way, is it a more general way to cancel the upstream `CompletableFuture` or `Publisher`.
I don't see the several callbacks are maintainable. I am also concerned about performance degradation.
How about introducing a new API to `ResponseConverterFunction`? I'm not sure 100% that this is the right decision. 
```java
public interface ResponseConverterFunction {
    @Deprecate
    @CheckReturnValue
    HttpResponse convertResponse(ServiceRequestContext ctx,
                                 ResponseHeaders headers,
                                 @Nullable Object result,
                                 HttpHeaders trailers) throws Exception;

    // Add new API for supporting cancellation
    // `whenCancelled` will be injected from AnnotatedService
    @CheckReturnValue
    default HttpResponse convertResponse(
            ServiceRequestContext ctx, ResponseHeaders headers, @Nullable Object result,
            HttpHeaders trailers, CompletableFuture<Void> whenCancelled) throws Exception {
        return convertResponse(ctx, headers, result, trailers);
    }
}

// In ObjectCollectingUtil.java
// Use whenCancelled for cancel upstream publisher
public static CompletableFuture<Object> collectFrom(Publisher<?> publisher, CompletableFuture<Void> whenCancelled) {
    ...
}
```

Reply:
Otherwise, we can use `ServiceRequestContext.whenRequestCancelling()` rather than introducing new API.

Reply:
> is it a more general way to cancel the upstream CompletableFuture or Publisher.

Agree that it would be nice if we have a more general way to back-propagate cancellation.

There are 2 cases that we tried to fix in this PR.
* Converting `CompletableFuture` to `Publisher(HttpResponse)` - In `AnnotatedService` and `ResponseConversionUtil`.
  - +1 for using `ctx.whenCancelling()` or `HttpResponse.whenComplete()` rather than extending the `ResponseConverterFunction` API. But I doubt `ResponseConverterFunction` is the right place to put cancellation propagation logic since it won't be called until the future that we want to cancel gets completed.
  - (By the way, thanks for the insight! I didn't notice `ctx.whenRequestCancelling()` API 😆)

* Converting `Publisher<?>` to `CompletableFuture<?>`
  - Maybe this will giv
  - IMHO, it is as good as it is now, since we will almost always use `ObjectCollectingUtil.collectFrom(...)` in this case, and it already propagates cancellation to publisher if this PR is merged.

> I am also concerned about performance degradation.

Currently, 2~3 (depending on the return type) callbacks will be triggered for every request which could be non-trivial. I can't think of a good way to improve this in case of `FUTURE` or `COROUTINE` return type, but maybe there's a room for `Publisher` return type if we introduce a new method in `ObjectCollectinUtil` which returns another publisher after collecting elements from the source, just like `Mono.collect(...)` whose implementation doesn't rely on `CompletableFuture`.

Reply:
> Collecting Publisher<?> to CompletableFuture<?>

IIUC, if `AbstractCollectingSubscriber` adds the cancellation callback on `ctx.whenRequestCancelling()` instead of the collection `future`, 
we could remove the lines 561 to 566 (https://github.com/line/armeria/pull/3691/files#diff-a5b50e72b0031befe6de444cf1fc3d47c573cd73a65312068dd7ba99d440f8b1R561-R566) in `AnnotatedService`.

Reply:
@ikhoon Thanks for the idea, I've sent a new commit to check if I understood your points correctly, PTAL 🙏.
https://github.com/line/armeria/pull/3691/commits/5563f7f37cefd0a306307164f1b3645fb78ad4b9

Reply:
Perfect! That is what I imagined. 💯